Sources/HarmonizeSemantics/GrammarComponents/Provider/BodyProviding.swift

```diff
@@ -18,7 +18,76 @@
 
 /// A protocol that represents declarations capable of providing a body.
-public protocol BodyProviding {
+public protocol BodyProviding: FunctionCallsProviding {
     /// The body of the declaration, if any.
     var body: Body? { get }
 }
+
+extension BodyProviding {
+    public var functionCalls: [FunctionCall] {
+        body?.functionCalls ?? []
+    }
+    
+    // MARK: - Checks for Assignments
+        
+    /// Checks if the body contains an assignment to a variable with the specified name.
+    ///
+    /// - Parameter variableName: The name of the variable to check for.
+    /// - Returns: `true` if the body contains an assignment to the variable, otherwise `false`.
+    public func assigns(to variableName: String) -> Bool {
+        body?.assignments.contains { $0.target == variableName } ?? false
+    }
+    
+    /// Retrieves all assignments to the specified variable name.
+    ///
+    /// - Parameter variableName: The name of the variable.
+    /// - Returns: An array of assignments to the specified variable, or an empty array if none exist.
+    public func assignments(to variableName: String) -> [Assignment] {
+        body?.assignments.filter { $0.target == variableName } ?? []
+    }
+    
+    /// Checks if the body contains any assignments.
+    ///
+    /// - Returns: `true` if the body contains assignments, otherwise `false`.
+    public func hasAssignments() -> Bool {
+        body?.assignments.isEmpty == false
+    }
+        
+    /// Checks if the body contains any `if` statements.
+    ///
+    /// - Returns: `true` if the body contains `if` statements, otherwise `false`.
+    public func hasIfStatements() -> Bool {
+        body?.ifs.isEmpty == false
+    }
+    
+    /// Retrieves all `if` statements in the body.
+    ///
+    /// - Returns: An array of `If` objects representing all `if` statements, or an empty array if none exist.
+    public func ifStatements() -> [If] {
+        body?.ifs ?? []
+    }
+        
+    /// Checks if the body contains any `switch` statements.
+    ///
+    /// - Returns: `true` if the body contains `switch` statements, otherwise `false`.
+    public func hasSwitchStatements() -> Bool {
+        body?.switches.isEmpty == false
+    }
+    
+    /// Retrieves all `switch` statements in the body.
+    ///
+    /// - Returns: An array of `Switch` objects representing all `switch` statements, or an empty array if none exist.
+    public func switchStatements() -> [Switch] {
+        body?.switches ?? []
+    }
+    
+    /// Checks if the body is empty (contains no statements, assignments, or function calls).
+    ///
+    /// - Returns: `true` if the body is empty, otherwise `false`.
+    public func isEmptyBody() -> Bool {
+        let isEmpty = body?.statements.isEmpty ?? true
+        let hasNoAssignments = body?.assignments.isEmpty ?? true
+        let hasNoFunctionCalls = body?.functionCalls.isEmpty ?? true
+        return isEmpty && hasNoAssignments && hasNoFunctionCalls
+    }
```

Comment:
This seems a bit error-prone, unless we're certain that we have captured all cases, right? Why don't we just check if the description of the body is empty? Wouldn't that work?

Reply:
Statements are supposed to capture all possibles cases we don't support, but I agree, this is error-prone and I think your solution is better, it is failsafe.